### PR TITLE
[feature] add support for sequence feature

### PR DESCRIPTION
**Pull request by lgqfhwy**
add support for sequence feature. config in samples/model_config/din_on_sequence_feature_taobao.config

### CODE REVIEW

File: easy_rec/python/layers/seq_input_layer.py [hunk 1]
```diff
@@ -22,43 +22,54 @@ def __init__(self, feature_configs, feature_groups_config):
...
-            x.hist_seq]._get_sequence_dense_tensor(builder)
+    def _seq_embed_name(input_name):
+      input_name = input_name.split(':')[0]
+      input_name = input_name.split('/')[:2]
+      return 'sequence_feature/' + '/'.join(input_name)
 
-      key_features.append(key_tensor)
-      hist_emb_features.append(hist_embedding)
-      hist_seqlen_features.append(hist_seq_len)
+    with tf.variable_scope(group_name, reuse=tf.AUTO_REUSE):
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply easy_rec/python/layers/seq_input_layer.py]: done

File: samples/model_config/din_on_sequence_feature_taobao.config [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    feature_names: 'adgroup_id'
+    feature_names: 'cate_id'
+    feature_names: 'campaign_id'
+    feature_names: 'customer'
+    feature_names: 'brand'
+    feature_names: 'price'
+    feature_names: 'pid'
+    wide_deep: DEEP
+  }
+  seq_att_groups: {
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply samples/model_config/din_on_sequence_feature_taobao.config:255]: now, add sequence_feature for all rank model

File: samples/model_config/din_on_sequence_feature_taobao.config [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+}
+feature_configs : {
+  input_names: 'new_user_class_level'
+  feature_type: IdFeature
+  embedding_dim: 16
+  hash_bucket_size: 10
+}
+feature_configs : {
+   input_names: 'tag_category_list'
+   feature_type: SequenceFeature
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply samples/model_config/din_on_sequence_feature_taobao.config:212]: done

File: easy_rec/python/layers/seq_input_layer.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 from easy_rec.python.compat.feature_column import feature_column
 from easy_rec.python.feature_column.feature_column import FeatureColumnParser
 from easy_rec.python.protos.feature_config_pb2 import WideOrDeep
-
+import logging
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply easy_rec/python/layers/seq_input_layer.py]: done

File: samples/model_config/din_on_sequence_feature_taobao.config (hunk 2 above)
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply samples/model_config/din_on_sequence_feature_taobao.config:255]: done

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]
add new test configs to train_eval_test

File: easy_rec/python/layers/seq_input_layer.py [hunk 5]
```diff
@@ -22,43 +24,60 @@ def __init__(self, feature_configs, feature_groups_config):
...
-          'hist_seq_len': hist_seqlen_features[0]
-      }
-    else:
-      features = {
-          'key': key_features[0],
-          'hist_seq_emb': hist_emb_features[0],
-          'hist_seq_len': hist_seqlen_features[0]
-      }
+    hist_tensors = []
+    for hist_seq in hist_seqs:
```
Comment (tiankongdeguiji) [inline easy_rec/python/layers/seq_input_layer.py:55]: `for x in feature_dict.seq_att_map` for forward compatbility

File: easy_rec/python/model/esmm.py [hunk 6]
```diff
@@ -34,12 +34,21 @@ def __init__(self,
...
       logging.info('group_num: {0}'.format(self._group_num))
+      exclude_group_names = []
       for group_id in range(self._group_num):
         group = self._model_config.groups[group_id]
-        group_feature, _ = self._input_layer(self._feature_dict, group.input)
+        if group.input != 'user':
+          exclude_group_names.append(group.input)
+      for group_id in range(self._group_num):
+        group = self._model_config.groups[group_id]
+        if group.input == 'user':
```
[the remark this answers is not in the document]
Comment (tiankongdeguiji) [reply easy_rec/python/model/esmm.py:44]: combine sequence_features and features into one group.

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

File: easy_rec/python/layers/seq_input_layer.py (hunk 5 above)
Comment (lgqfhwy) [reply easy_rec/python/layers/seq_input_layer.py]: done

File: easy_rec/python/model/esmm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply easy_rec/python/model/esmm.py]: done

File: easy_rec/python/layers/input_layer.py [hunk 7]
```diff
@@ -49,7 +61,111 @@ def __init__(self,
...
+      logging.info(
+          'seq_dnn not set in seq_att_groups, will use default settings')
+      from easy_rec.python.protos.dnn_pb2 import DNN
+      seq_dnn_config = DNN()
+      seq_dnn_config.hidden_units.extend([128, 64, 32, 1])
+    seq_fea = self.din(seq_dnn_config, seq_features, name='seq_dnn')
+    return seq_fea
+
+  def __call__(self, features, *group_names, **kwargs):
+    is_combine = True
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply easy_rec/python/layers/input_layer.py]: done

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

File: easy_rec/python/utils/convert_rtp_fg.py [hunk 8]
```diff
@@ -246,6 +282,22 @@ def convert_rtp_fg(rtp_fg,
                                                         embedding_dim,
                                                         incol_separator)
 
+  new_rtp_features = []
+  for feature in rtp_features:
+    if 'feature_name' in feature:
+      new_rtp_features.append(feature)
+    elif 'sequence_name' in feature:
+      print('hello feature = ', feature)
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply easy_rec/python/utils/convert_rtp_fg.py]: done

File: docs/source/feature/feature.rst [hunk 9]
```diff
@@ -241,6 +241,39 @@ Sequense类特征格式一般为“XX\|XX\|XX”，如用户行为序列特征
...
+    feature_names: 'cms_segid'
+    feature_names: 'cms_group_id'
+    feature_names: 'age_level'
+    feature_names: 'pvalue_level'
+    feature_names: 'shopping_level'
+    feature_names: 'occupation'
+    feature_names: 'new_user_class_level'
+    wide_deep:DEEP
+    sequence_features: {
+    group_name: "seq_fea"
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply docs/source/feature/feature.rst:261]: 需要的，因为在对于 config 的处理中，我们是分层来处理，上一层不用去关心下一层的事情。而在 seq_input_layer 这一层中，它就专心只处理好 SeqAttGroupConfig 这一层的事情，这一层的处理除了这种包含的情况，更重要是独立一个 feature group 的情况。所以为了与独立一个 feature group 的情况一致，需要加上这个 group name.

File: docs/source/feature/feature.rst [hunk 10]
```diff
@@ -241,6 +241,39 @@ Sequense类特征格式一般为“XX\|XX\|XX”，如用户行为序列特征
 -  hash\_bucket\_size: 同离散值特征
 -  NOTE：SequenceFeature一般用在DIN算法或者BST算法里面。
 
+现在，已经在所有模型中均支持上了序列特征。具体原理是先将序列特征过 Target Attention 模型 (DIN), 然后再
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply docs/source/feature/feature.rst:244]: done

File: easy_rec/python/layers/input_layer.py [hunk 11]
```diff
@@ -49,6 +66,62 @@ def __init__(self,
   def has_group(self, group_name):
     return group_name in self._feature_groups
 
+  def din(self, dnn_config, deep_fea, name):
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply easy_rec/python/layers/input_layer.py:69]: done

File: easy_rec/python/layers/seq_input_layer.py [hunk 12]
```diff
@@ -22,43 +24,117 @@ def __init__(self, feature_configs, feature_groups_config):
...
+                _seq_embed_summary_name(hist_seq_len.name), hist_seq_len)
+
+    features = {
+        'key': tf.concat(key_tensors, axis=-1),
+        'hist_seq_emb': tf.concat([x[0] for x in hist_tensors], axis=-1),
+        'hist_seq_len': hist_tensors[0][1]
+    }
+    return features
+
+  def pure_sequence_feature_call(self,
```
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply easy_rec/python/layers/seq_input_layer.py:76]: done.

File: docs/source/feature/feature.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (lgqfhwy) [reply docs/source/feature/feature.rst]: done

**Review by tiankongdeguiji — APPROVED** [review]
LGTM

**Pull request merged**